Update CMake script

jibsen opened the pull request:
Okay, as mentioned in #143 I've taken a shot at updating the CMake script to try to address some of the issues raised there. Specifically:

  - Add install target
  - Version is only set once
  - The default build type now uses CFLAGS etc.

I also improved support for using Zopfli as a subproject from other CMake based projects.

Some issues/things to discuss that remain:

  - I added a workaround for the issue about symbol export when building as a DLL with MSVC, it only works for recent versions of CMake. Actually fixing this would require adding macros to the public headers to optionally export symbols, and I don't know if the devs are interested in this (see for instance [this description](https://gcc.gnu.org/wiki/Visibility)).
  - libm is still included by default on UNIX, somebody with more knowledge about this will have to fix it if needed.
  - Leaving the default build type alone allows building using CFLAGS etc., but means the default build will be unoptimized, perhaps the docs should suggest building with -DCMAKE_BUILD_TYPE=Release?

jibsen commented:
Digging around the libm issue a bit, I found a number of projects that test specifically for BeOS and Haiku before adding the library -- I've added this to the PR.

If it turns out this is not enough, we could explicitly check if we can link in log() without including libm;

~~~.cmake
include(CheckFunctionExists)
check_function_exists(log zopfli_has_log)
if(NOT zopfli_has_log)
  target_link_libraries(libzopfli PRIVATE m)
endif()
~~~

lvandeve commented:
I built zopfli with the new CMakeLists.txt for this pull request, and also with the current CMakeLists.txt before this pull request.

The zopfli binary of before appears to run 3x slower after than before this pull request. Might some compiler flags be different? I don't see any flags like O3 in either of them so I'm not sure how to tell.

Do you measure any speed difference between the two on your system?

I built like this in both cases, on Linux:

cmake . -DBUILD_SHARED_LIBS=ON && make VERBOSE=1

Ideally it should have the same speed as before

jibsen commented:
Yes, as noted above, I made it so the build type is not forced to Release if unset. This means you have to specify the Release build if you want it:
~~~.sh
cmake . -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=ON && make VERBOSE=1
~~~
The reason for this change was the issues @hartwork had in #143 with being unable to override the build flags with CFLAGS/CXXFLAGS.

If you prefer having the build default to Release and using one of the workarounds from that thread to support CFLAGS, I'll change that back.

lvandeve commented:
What do open source projects normally do, do people expect release mode or debug mode by default? Is the override of CFLAGS an issue in other open source projects and how do they deal with it?

As a data point, personally when I build (make) a project created by somebody else I assume that by default it builds a fast release mode version. What is your experience with this?

jibsen commented:
> What do open source projects normally do, do people expect release mode or debug mode by default? Is the override of CFLAGS an issue in other open source projects and how do they deal with it?

I am afraid it varies. I wish there were some clear best practices for things like this, but in general I think as long as you clearly state in the documentation how to build it, people can work it out.

Regarding the CFLAGS issue, I don't have any experience with this myself, I believe it is something package maintainers use? but I was hoping some of the people who have raised issues around it might chime in.

> As a data point, personally when I build (make) a project created by somebody else I assume that by default it builds a fast release mode version. What is your experience with this?

I would agree that If a project has a bare Makefile, I assume it will build a release version. If it has a CMake script, I assume I can choose which build type I'd like when running CMake (for a single-configuration generator). This also works if we default to Release of course, the question is if it is better or worse to change the default behavior of CMake?

I can see arguments in favor of both, personally I am gravitating towards it being best to try to avoid changing default behavior and global variables unless your build depends on it -- especially if your project might be used as a subproject. But of course you risk people building the latest version without reading the docs on setting a build type and going "Zopfli is slow!".

Here are some more datapoints:

  - Brotli does not set build type and the [README](https://github.com/google/brotli#cmake) desribes how to build Release
  - Zlib does not appear to set build type in their [CMake script](https://github.com/madler/zlib/blob/master/CMakeLists.txt)
  - zstd forces the build type to Release in their [CMake script](https://github.com/facebook/zstd/blob/dev/build/cmake/CMakeLists.txt)
  - LLVM defaults to Debug if no build type is set in their [CMake script](https://github.com/llvm-mirror/llvm/blob/master/CMakeLists.txt)
  - MySQL defaults to RelWithDebInfo if no build type is set in their [CMake script](https://github.com/mysql/mysql-server/blob/8.0/CMakeLists.txt)

As you can see, there are examples of pretty much every possible choice -- the question is what do you prefer? (and what works best for your users of course).

hartwork commented:
>  Is the override of CFLAGS an issue in other open source projects and how do they deal with it?

As a distro packager you want the package to build 100% your distro-specific flags without need to patch any files. With a source-based distro like Gentoo, flags are user specific even.

hartwork commented on CMakeLists.txt:
```diff
@@ -1,17 +1,54 @@
 project(Zopfli)
 
-option(BUILD_SHARED_LIBS "Build Zopfli with shared libraries" OFF)
+# Check if being built standalone
```
Standalone build means to build in the source tree without a separate build directory, right?
I still haven't fully figure it out so please take my vote on making this check and it's use more obvious. Just my 2 cents.

jibsen replied:
In CMake terminology, a build in the source tree is called an in-source build, and a build in a separate build directory an out-of-source build.

What I mean by standalone is that Zopfli is being built as the top-level (only) CMake project, as opposed to as a subproject of some other CMake based build.

An example of the latter would be [pngwolf-zopfli](https://github.com/jibsen/pngwolf-zopfli), which includes Zopfli as a git submodule. If Zopfli works as a CMake subproject, I would be able to do build and link it with
~~~.cmake
add_subdirectory(zopfli)
target_link_libraries(pngwolf PRIVATE libzopflipng)
~~~
rather than list all the source files inside the Zopfli directory.

Maybe the comment should be
~~~.cmake
# Check if being built standalone or as subproject
~~~

hartwork replied:
That comment would help :+1:

hartwork commented on CMakeLists.txt:
```diff
@@ -1,17 +1,54 @@
+if(DEFINED BUILD_SHARED_LIBS)
+  set(zopfli_shared_default ${BUILD_SHARED_LIBS})
 endif()
+option(ZOPFLI_BUILD_SHARED "Build Zopfli with shared libraries" ${zopfli_shared_default})
```
These lines look a bit as if `BUILD_SHARED_LIBS` and `ZOPFLI_BUILD_SHARED` could both be used to flip the same bit from the outside. Are both needed or can this be simplified to a single variable?

jibsen replied:
They could indeed. The reason for the separate variable is again the potential use as a subproject.

BUILD_SHARED_LIBS is a global variable that causes all libraries that are not specifically static to be built shared.

If Zopfli is being built as a subproject, we should not change the state of this variable, potentially changing how other parts of the parent project are being built. On the other hand, unless told otherwise, we should respect the value of this variable if it has been set.

The logic here is an attempt to support this, and also to allow situations like building Zopfli as a static library as part of a build where BUILD_SHARED_LIBS is ON, or vice versa.

So, as you noted, this extra variable is redundant for building Zopfli on its own, but hopefully useful when built as a subproject.

Another option is to use a normal variable rather than an option for this, for instance BROTLI_BUNDLED_MODE in [Brotli](https://github.com/google/brotli/blob/master/CMakeLists.txt).

hartwork replied:
Thanks for the explanation.

hartwork commented on CMakeLists.txt:
```diff
@@ -1,17 +1,54 @@
-if(MSVC)
-  add_definitions(/D_CRT_SECURE_NO_WARNINGS)
+# If standalone or shared subproject, default to building install
+if(zopfli_standalone OR ZOPFLI_BUILD_SHARED)
```
I'm unsure why that guard was chosen for enabling the install feature. I'd have expected something like "enabled unless on WIN32" or so. What is the idea?

jibsen replied:
If we are building Zopfli only (standalone), we want to enable install.

If Zopfli is being built as a subproject of some other project, and as static libraries, then we will most likely be linked into the parent project, and should not add our installation files to the parent project (unless asked to).

If however Zopfli is being built as a subproject as shared libraries, it will most likely need to be installed alongside the parent project for it to work.

I don't know, perhaps this is too much logic to put into the CMake script to try and cover obscure use cases?

hartwork replied:
Make more sense to me now, thanks.

hartwork commented on CMakeLists.txt:
```diff
@@ -42,47 +79,81 @@ set (lodepng_src
+add_library(libzopflipng ${zopfli_library_type}
   ${zopflipnglib_src}
   ${lodepng_src}
   $<TARGET_OBJECTS:zopflilib_obj>
```
My vote for using `libzopfli` instead here to save RAM and diskspace. At least an option to avoid such duplication would be nice. Just my 2 cents.

hartwork commented on CMakeLists.txt:
```diff
@@ -42,47 +79,81 @@ set (lodepng_src
+    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+  )
+  install(FILES include/zopfli.h include/zopflipng_lib.h
```
That line seems to cause an install error:
```console
# make DESTDIR="$PWD/DESTDIR" install
[..]
CMake Error at cmake_install.cmake:171 (file):
  file INSTALL cannot find
  "/tmp/tmp.EK9KAgZMcc/jibsen-zopfli--update-cmake/include/zopfli.h".

# find -name zopfli.h
./src/zopfli/zopfli.h
```

jibsen replied:
Ah, I copied the install section from an older version of the CMake script, which used the previous directory structure. I will update that, thanks!

hartwork replied:
Works better now, thanks!

```
# make DESTDIR="$PWD/DESTDIR" install
[..]
# find DESTDIR/ -not -type d | sort
DESTDIR/usr/local/bin/zopfli
DESTDIR/usr/local/bin/zopflipng
DESTDIR/usr/local/include/zopfli.h
DESTDIR/usr/local/include/zopflipng_lib.h
DESTDIR/usr/local/lib64/libzopflipng.so
DESTDIR/usr/local/lib64/libzopflipng.so.1
DESTDIR/usr/local/lib64/libzopflipng.so.1.0.2
DESTDIR/usr/local/lib64/libzopfli.so
DESTDIR/usr/local/lib64/libzopfli.so.1
DESTDIR/usr/local/lib64/libzopfli.so.1.0.2
```

jibsen commented:
Great feedback @hartwork, thanks!

jibsen commented:
Updated some of the comments to try to better explain the idea behind the options.

I switched to libzopflipng linking with libzopfli since it simplified the build somewhat. We'll see if anyone complains.

Also by default set the build type to Release if empty, with an option (ZOPFLI_DEFAULT_RELEASE) to disable this behavior. If you want to have the empty build type so you can control the flags, something like this should work:
~~~.sh
CFLAGS='-O2' CXXFLAGS='-Os' LDFLAGS='-Wl,--as-needed' cmake -DBUILD_SHARED_LIBS=ON -DZOPFLI_DEFAULT_RELEASE=OFF .
~~~

jibsen commented:
The config file package support was the last thing I'd been meaning to add. It installs a few extra files that allow other CMake based projects to easily use an installed Zopfli with find_package().

lvandeve commented:
Seems to build optimized binaries again by default, good one to try, let's see if maintainers will like it. Thank you for the updates!
